src/System.Text.JsonLab.Serialization/System/Text/Json/Serialization/JsonConverter.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Text.Json.Serialization
+{
+    public static partial class JsonConverter
+    {
+        private static readonly JsonConverterSettings s_default_options = new JsonConverterSettings();
+
+        private static bool ReadData(
+            ref Utf8JsonReader reader,
+            JsonConverterSettings options,
+            Type returnType,
+            ref JsonObjectState current,
+            ref List<JsonObjectState> previous,
+            ref int arrayIndex)
+        {
+            while (reader.Read())
+            {
+                JsonTokenType tokenType = reader.TokenType;
```

Comment:
nit: consider using switch?

Reply:
I'll measure perf; switch will change IL generation from elseif to an "optimized" version that may or may not be faster since currently with elseif the ordering is from common to less common.